source/opt/workaround_image_operands.cpp

```diff
@@ -0,0 +1,160 @@
+          opt::Instruction* sampled_image_usage, uint32_t idx) {
+        bool has_depth = false;
+        if (sampled_image_usage->opcode() == SpvOpImageDrefGather ||
+            sampled_image_usage->opcode() == SpvOpImageSampleDrefImplicitLod ||
+            sampled_image_usage->opcode() == SpvOpImageSampleDrefExplicitLod ||
+            sampled_image_usage->opcode() ==
+                SpvOpImageSampleProjDrefImplicitLod ||
+            sampled_image_usage->opcode() ==
+                SpvOpImageSampleProjDrefExplicitLod ||
+            sampled_image_usage->opcode() ==
+                SpvOpImageSparseSampleDrefImplicitLod ||
+            sampled_image_usage->opcode() ==
+                SpvOpImageSparseSampleDrefExplicitLod ||
+            sampled_image_usage->opcode() ==
+                SpvOpImageSparseSampleProjDrefImplicitLod ||
+            sampled_image_usage->opcode() ==
+                SpvOpImageSparseSampleProjDrefExplicitLod ||
+            sampled_image_usage->opcode() == SpvOpImageSparseDrefGather) {
+          has_depth = true;
+        }
+        if (!has_depth) {
+          return;
+        }
+        analysis::Image* image_type = context()
+                                          ->get_type_mgr()
+                                          ->GetType(img_type->result_id())
+                                          ->AsImage();
+        analysis::Type* t = context()->get_type_mgr()->GetType(
+            context()->get_type_mgr()->GetId(image_type->sampled_type()));
+        analysis::Image img(
+            t, image_type->dim(), has_depth ? 1 : 0, image_type->is_arrayed(),
+            image_type->is_multisampled(), image_type->sampled(),
+            image_type->format(), image_type->access_qualifier());
+        for (auto& dec : image_type->decorations()) {
+          img.AddDecoration(std::vector<uint32_t>(dec));
+        }
+        uint32_t new_image =
+            context()->get_type_mgr()->GetTypeInstruction(&img);
+
+        analysis::SampledImage sampled_img(
+            context()->get_type_mgr()->GetType(new_image));
+        for (auto& dec : sampled_image_type->decorations()) {
+          sampled_img.AddDecoration(std::vector<uint32_t>(dec));
+        }
+
+        uint32_t sampled_image_type =
+            context()->get_type_mgr()->GetTypeInstruction(&sampled_img);
+
+        auto new_sampled_image = op_sampled_image->Clone(context());
+        new_sampled_image->SetResultType(sampled_image_type);
+
+        new_sampled_image->InsertBefore(op_sampled_image);
+        new_sampled_image->SetResultId(context()->TakeNextId());
+
+        sampled_image_usage->SetOperand(idx, {new_sampled_image->result_id()});
+        context()->AnalyzeUses(sampled_image_usage);
+        context()->AnalyzeDefUse(new_sampled_image);
+      });
+}
+
+void WorkaroundImageOperands::WalkImageUse(
+    Instruction* inst, Instruction* img_type,
+    analysis::DefUseManager* def_use_mgr) {
```

Comment:
You probably don't need to pass the def_use_mgr around.  It is part of the context, which is available anywhere in the pass.